File: tw128.py
"""TW128 reference implementation.

TW128 is the 128-bit TreeWrap instantiation using Keccak-p[1600,12], a
256-bit capacity, 256-bit keys, 256-bit nonces, and 256-bit tags.
"""

from __future__ import annotations

import hmac

from duplex import MAX_BYTE_ALIGNED_SIGMA_BYTES, Duplex

KEY_BYTES = 32
NONCE_BYTES = 32
TAG_BYTES = 32
CHUNK_BYTES = 8183
MAX_MESSAGE_BYTES = (1 << 64) * CHUNK_BYTES
RHO_BYTES = MAX_BYTE_ALIGNED_SIGMA_BYTES
SUFFIX_BITS = 4

ROOT_PREFIX = b"TW128R"
LEAF_PREFIX = b"TW128L"

INIT_LAST = 0x0C
AD_MORE = 0x09
AD_LAST = 0x0D
MSG_MORE = 0x0A
MSG_LAST = 0x0E
AGG_MORE = 0x0B
AGG_LAST = 0x0F


class InvalidTag(Exception):
    """Raised when TW128 authentication fails."""


def encrypt(
    key: bytes, nonce: bytes, plaintext: bytes, associated_data: bytes
) -> bytes:
    """Return ciphertext || 32-byte root tag."""
    ciphertext, tag = encrypt_detached(key, nonce, plaintext, associated_data)
    return ciphertext + tag


def decrypt(
    key: bytes,
    nonce: bytes,
    ciphertext_and_tag: bytes,
    associated_data: bytes,
) -> bytes:
    """Verify ciphertext || tag and return plaintext, or raise InvalidTag."""
    ciphertext_and_tag = _as_bytes("ciphertext_and_tag", ciphertext_and_tag)
    if len(ciphertext_and_tag) < TAG_BYTES:
        raise InvalidTag("ciphertext is shorter than the TW128 tag")
    ciphertext = ciphertext_and_tag[:-TAG_BYTES]
    tag = ciphertext_and_tag[-TAG_BYTES:]
    return decrypt_detached(key, nonce, ciphertext, associated_data, tag)


def encrypt_detached(
    key: bytes,
    nonce: bytes,
    plaintext: bytes,
    associated_data: bytes,
) -> tuple[bytes, bytes]:
    """Return (ciphertext, root_tag)."""
    key, nonce, associated_data, plaintext = _validate_common(
        key, nonce, associated_data, plaintext
    )
    if len(plaintext) > MAX_MESSAGE_BYTES:
        raise ValueError(f"plaintext must be at most {MAX_MESSAGE_BYTES} bytes")

    root_plaintext = plaintext[:CHUNK_BYTES]
    leaf_plaintext = plaintext[CHUNK_BYTES:]

    root, keystream = _new_root(
        key, nonce, associated_data, _first_block_len(root_plaintext)
    )

    if not leaf_plaintext:
        # Single chunk (no leaves): elide the aggregation phase. The closing
        # root MSG_LAST call emits the root tag directly, mirroring a leaf.
        root_ciphertext, root_tag = _encrypt_overwrite(
            root, root_plaintext, keystream, TAG_BYTES
        )
        return root_ciphertext, root_tag

    root_ciphertext, _ = _encrypt_overwrite(root, root_plaintext, keystream, 0)

    ciphertext_parts = [root_ciphertext]
    leaf_tags = []
    for chunk_id, chunk in enumerate(_chunks(leaf_plaintext, CHUNK_BYTES), start=1):
        leaf, leaf_keystream = _new_leaf(key, nonce, chunk_id, _first_block_len(chunk))
        leaf_ciphertext, leaf_tag = _encrypt_overwrite(
            leaf, chunk, leaf_keystream, TAG_BYTES
        )
        ciphertext_parts.append(leaf_ciphertext)
        leaf_tags.append(leaf_tag)

    root_tag = _absorb_leaf_tags(root, leaf_tags)
    return b"".join(ciphertext_parts), root_tag


def decrypt_detached(
    key: bytes,
    nonce: bytes,
    ciphertext: bytes,
    associated_data: bytes,
    tag: bytes,
) -> bytes:
    """Verify detached tag and return plaintext, or raise InvalidTag."""
    key, nonce, associated_data, ciphertext = _validate_common(
        key, nonce, associated_data, ciphertext
    )
    if len(ciphertext) > MAX_MESSAGE_BYTES:
        raise ValueError(f"ciphertext must be at most {MAX_MESSAGE_BYTES} bytes")
    tag = _as_bytes("tag", tag)
    if len(tag) != TAG_BYTES:
        raise InvalidTag(f"tag must be {TAG_BYTES} bytes")

    root_ciphertext = ciphertext[:CHUNK_BYTES]
    leaf_ciphertext = ciphertext[CHUNK_BYTES:]

    root, keystream = _new_root(
        key, nonce, associated_data, _first_block_len(root_ciphertext)
    )

    if not leaf_ciphertext:
        # Single chunk (no leaves): elide the aggregation phase. The closing
        # root MSG_LAST call emits the root tag directly, mirroring a leaf.
        root_plaintext, expected_tag = _decrypt_overwrite(
            root, root_ciphertext, keystream, TAG_BYTES
        )
        if not hmac.compare_digest(expected_tag, tag):
            raise InvalidTag("TW128 authentication failed")
        return root_plaintext

    root_plaintext, _ = _decrypt_overwrite(root, root_ciphertext, keystream, 0)

    plaintext_parts = [root_plaintext]
    leaf_tags = []
    for chunk_id, chunk in enumerate(_chunks(leaf_ciphertext, CHUNK_BYTES), start=1):
        leaf, leaf_keystream = _new_leaf(key, nonce, chunk_id, _first_block_len(chunk))
        leaf_plaintext, leaf_tag = _decrypt_overwrite(
            leaf, chunk, leaf_keystream, TAG_BYTES
        )
        plaintext_parts.append(leaf_plaintext)
        leaf_tags.append(leaf_tag)

    expected_tag = _absorb_leaf_tags(root, leaf_tags)
    if not hmac.compare_digest(expected_tag, tag):
        raise InvalidTag("TW128 authentication failed")
    return b"".join(plaintext_parts)


def _new_root(
    key: bytes, nonce: bytes, associated_data: bytes, first_out_len: int
) -> tuple[Duplex, bytes]:
    duplex = Duplex()
    init = ROOT_PREFIX + key + nonce
    if associated_data:
        # Associated-data phase: the init call emits no keystream, and the
        # closing AD_LAST block produces the first chunk-0 keystream block.
        duplex.duplexing_bits(init, INIT_LAST, SUFFIX_BITS, 0)
        keystream = _absorb_ad(duplex, associated_data, first_out_len)
    else:
        # Empty associated data: elide the AD phase entirely. The init call
        # itself emits the first chunk-0 keystream block, mirroring leaf init.
        keystream = duplex.duplexing_bits(init, INIT_LAST, SUFFIX_BITS, first_out_len)
    return duplex, keystream


def _new_leaf(
    key: bytes, nonce: bytes, chunk_id: int, first_out_len: int
) -> tuple[Duplex, bytes]:
    duplex = Duplex()
    init = LEAF_PREFIX + key + nonce + _le_u64(chunk_id)
    keystream = duplex.duplexing_bits(init, INIT_LAST, SUFFIX_BITS, first_out_len)
    return duplex, keystream


def _absorb_ad(duplex: Duplex, associated_data: bytes, first_out_len: int) -> bytes:
    blocks = _split_at_least_one(associated_data, RHO_BYTES)
    for block in blocks[:-1]:
        duplex.duplexing_bits(block, AD_MORE, SUFFIX_BITS, 0)
    return duplex.duplexing_bits(blocks[-1], AD_LAST, SUFFIX_BITS, first_out_len)


def _encrypt_overwrite(
    duplex: Duplex,
    plaintext: bytes,
    keystream: bytes,
    final_out_len: int,
) -> tuple[bytes, bytes]:
    blocks = _split_at_least_one(plaintext, RHO_BYTES)
    ciphertext = bytearray()
    current_keystream = keystream
    for index, block in enumerate(blocks):
        if len(current_keystream) != len(block):
            raise ValueError("keystream length does not match plaintext block length")
        encrypted = _xor_bytes(block, current_keystream)
        ciphertext.extend(encrypted)
        last = index == len(blocks) - 1
        next_len = final_out_len if last else len(blocks[index + 1])
        suffix = MSG_LAST if last else MSG_MORE
        current_keystream = duplex.duplexing_bits(
            encrypted, suffix, SUFFIX_BITS, next_len
        )
    return bytes(ciphertext), current_keystream


def _decrypt_overwrite(
    duplex: Duplex,
    ciphertext: bytes,
    keystream: bytes,
    final_out_len: int,
) -> tuple[bytes, bytes]:
    blocks = _split_at_least_one(ciphertext, RHO_BYTES)
    plaintext = bytearray()
    current_keystream = keystream
    for index, block in enumerate(blocks):
        if len(current_keystream) != len(block):
            raise ValueError("keystream length does not match ciphertext block length")
        decrypted = _xor_bytes(block, current_keystream)
        plaintext.extend(decrypted)
        last = index == len(blocks) - 1
        next_len = final_out_len if last else len(blocks[index + 1])
        suffix = MSG_LAST if last else MSG_MORE
        current_keystream = duplex.duplexing_bits(block, suffix, SUFFIX_BITS, next_len)
    return bytes(plaintext), current_keystream


def _absorb_leaf_tags(duplex: Duplex, leaf_tags: list[bytes]) -> bytes:
    transcript = b"".join(leaf_tags) + _le_u64(len(leaf_tags))
    blocks = _split_at_least_one(transcript, RHO_BYTES)
    for block in blocks[:-1]:
        duplex.duplexing_bits(block, AGG_MORE, SUFFIX_BITS, 0)
    return duplex.duplexing_bits(blocks[-1], AGG_LAST, SUFFIX_BITS, TAG_BYTES)


def _validate_common(
    key: bytes,
    nonce: bytes,
    associated_data: bytes,
    data: bytes,
) -> tuple[bytes, bytes, bytes, bytes]:
    key = _as_bytes("key", key)
    nonce = _as_bytes("nonce", nonce)
    associated_data = _as_bytes("associated_data", associated_data)
    data = _as_bytes("data", data)
    if len(key) != KEY_BYTES:
        raise ValueError(f"key must be {KEY_BYTES} bytes")
    if len(nonce) != NONCE_BYTES:
        raise ValueError(f"nonce must be {NONCE_BYTES} bytes")
    return key, nonce, associated_data, data


def _split_at_least_one(data: bytes, size: int) -> list[bytes]:
    if not data:
        return [b""]
    return [data[index : index + size] for index in range(0, len(data), size)]


def _chunks(data: bytes, size: int) -> list[bytes]:
    return [data[index : index + size] for index in range(0, len(data), size)]


def _first_block_len(data: bytes) -> int:
    return 0 if not data else min(len(data), RHO_BYTES)


def _xor_bytes(left: bytes, right: bytes) -> bytes:
    return bytes(a ^ b for a, b in zip(left, right, strict=True))


def _le_u64(value: int) -> bytes:
    if value < 0 or value >= 1 << 64:
        raise ValueError("value does not fit in le_u64")
    return value.to_bytes(8, "little")


def _as_bytes(name: str, value: bytes) -> bytes:
    if not isinstance(value, (bytes, bytearray, memoryview)):
        raise TypeError(f"{name} must be bytes-like")
    return bytes(value)

File: tw128/tw128.go
// Package tw128 implements TW128, a tree-parallel authenticated encryption algorithm based on keyed duplexes.
//
// WARNING: TreeWrap/TW128 is an unreviewed research construction. This package
// is hazmat code, not a production cryptographic library. It has not been
// audited, standardized, or hardened for deployment, and it should not be used
// to protect real data.
//
// This package exists to support the accompanying paper, experiments, test
// vectors, and benchmarks. Its API is the cipher.AEAD returned by New: Seal
// appends a TagSize-byte authentication tag and Open verifies it in constant
// time.
//
// The trunk (root) duplex handles associated-data absorption, encryption of
// chunk 0, absorption of the later hidden leaf tags, and final authentication
// tag extraction. Later chunks are processed by independent leaf transcripts
// under domain-separated chunk IDs.
//
// Each duplex is a strict BDPV11 duplex over Keccak-p[1600,12]: the root is
// initialized with "TW128R" || K || N and each leaf with "TW128L" || K || N ||
// le64(j). Every duplexing call absorbs at most rhoBytes of sigma followed by a
// 4-bit domain suffix and pad10*1, then permutes once.
package tw128

import (
	"encoding/binary"
)

const (
	// KeySize is the size of the key in bytes.
	KeySize = 32

	// NonceSize is the size of the nonce in bytes.
	NonceSize = 32

	// TagSize is the size of the authentication tag in bytes.
	TagSize = 32

	// ChunkSize is the size of each chunk in bytes. 8183 = 49 × rhoBytes (167),
	// so a chunk splits into exactly 49 full rho-blocks with no ragged tail.
	ChunkSize = 8183

	// leafTagSize is the size of each hidden leaf tag in bytes.
	leafTagSize = 32

	// prefixLen is the length of the role-distinguishing init prefix.
	prefixLen = 6

	// rootInitLen and leafInitLen are the lengths of the init sigma blocks.
	rootInitLen = prefixLen + KeySize + NonceSize
	leafInitLen = prefixLen + KeySize + NonceSize + 8
)

// Combined duplex suffix bytes. Each is the 4-bit domain suffix OR'd with the
// pad10*1 start bit (0x10); closeBlock XORs the trailing 0x80 pad bit.
const (
	initLast = 0x1C // 0x0C | 0x10
	adMore   = 0x19 // 0x09 | 0x10
	adLast   = 0x1D // 0x0D | 0x10
	msgMore  = 0x1A // 0x0A | 0x10
	msgLast  = 0x1E // 0x0E | 0x10
	aggMore  = 0x1B // 0x0B | 0x10
	aggLast  = 0x1F // 0x0F | 0x10
)

var (
	rootPrefix = [prefixLen]byte{'T', 'W', '1', '2', '8', 'R'}
	leafPrefix = [prefixLen]byte{'T', 'W', '1', '2', '8', 'L'}
)

// rootInit builds the root duplex init sigma: "TW128R" || key || nonce.
func rootInit(key, nonce []byte) [rootInitLen]byte {
	var p [rootInitLen]byte
	copy(p[:prefixLen], rootPrefix[:])
	copy(p[prefixLen:prefixLen+KeySize], key)
	copy(p[prefixLen+KeySize:], nonce)
	return p
}

// leafInit builds a leaf duplex init sigma: "TW128L" || key || nonce ||
// le64(chunkID).
func leafInit(key, nonce []byte, chunkID uint64) [leafInitLen]byte {
	var p [leafInitLen]byte
	copy(p[:prefixLen], leafPrefix[:])
	copy(p[prefixLen:prefixLen+KeySize], key)
	copy(p[prefixLen+KeySize:prefixLen+KeySize+NonceSize], nonce)
	binary.LittleEndian.PutUint64(p[prefixLen+KeySize+NonceSize:], chunkID)
	return p
}

// aggregator carries the state shared between the one-shot pipeline and the
// leaf-cascade kernels: the trunk duplex absorbing the leaf tags, the key and
// nonce for leaf inits, and the count of completed leaves. The trunk is held
// by value so the zero aggregator is valid and no field escapes to the heap.
type aggregator struct {
	key     [KeySize]byte
	nonce   [NonceSize]byte
	trunk   duplex // trunk (root) duplex state
	nLeaves uint64 // number of completed leaves
	decrypt bool
}

// crypt runs the one-shot TW128 pipeline over src into dst and returns the
// root tag. dst and src must be the same length and must overlap entirely or
// not at all. decrypt selects the SpongeWrap direction (the ciphertext is
// absorbed either way).
func crypt(key, nonce, ad, dst, src []byte, decrypt bool) [TagSize]byte {
	checkSize("key", key, KeySize)
	checkSize("nonce", nonce, NonceSize)

	var g aggregator
	copy(g.key[:], key)
	copy(g.nonce[:], nonce)
	g.decrypt = decrypt

	// Root init (INIT_LAST). When associated data is present, run the
	// associated-data phase; its closing AD_LAST block leaves the first
	// chunk-0 keystream block in the rate. When the associated data is empty,
	// the phase is elided: initWith already left the chunk-0 keystream block
	// in the rate, mirroring leaf init.
	p := rootInit(g.key[:], g.nonce[:])
	g.trunk.initWith(p[:])
	if len(ad) > 0 {
		g.trunk.absorbMore(ad, adMore)
		g.trunk.closeBlock(adLast)
	}

	// A message that fits in chunk 0 (including the empty message) is the
	// trunk message phase alone, closed with MSG_LAST, and the aggregation
	// phase is elided: the closing MSG_LAST block emits the root tag
	// directly, mirroring a leaf.
	if len(src) <= ChunkSize {
		g.trunk.bodyMore(dst, src, decrypt, msgMore)
		g.trunk.closeBlock(msgLast)
		var tag [TagSize]byte
		g.trunk.extractTag(&tag)
		return tag
	}

	// Partial lane-0 fusion: for a message with chunk 0 plus a ragged first leaf,
	// process their shared full MSG_MORE body blocks together where the platform
	// has a pair kernel, then finish their different final blocks separately.
	processedAll := false
	if len(src) < 2*ChunkSize {
		tailLen := len(src) - ChunkSize
		if decrypt {
			processedAll = decryptChunk0PartialFused(&g, src, dst, tailLen)
		} else {
			processedAll = encryptChunk0PartialFused(&g, src, dst, tailLen)
		}
	}

	if !processedAll {
		// Lane-0 fusion: chunk 0 has the same kernel-visible block schedule as a
		// leaf chunk and is contiguous with leaves 1..k-1 in the message, so on
		// platforms whose kernels hand the permutation state back, the trunk's
		// chunk-0 phase rides lane 0 of the first kernel call alongside up to
		// seven leaves, eliminating the serial chunk-0 pass. The fused call
		// absorbs the leaf tags it produced and advances g.nLeaves, and reports
		// how many chunks it consumed: k on amd64, 8 or 2 on arm64 (full x8
		// batch or one NEON pair), and 0 on platforms without a fused path,
		// where the trunk processes chunk 0 serially instead. processComplete
		// then continues the cascade with g.nLeaves pre-advanced.
		fusedChunks := 0
		if nComplete := (len(src) - ChunkSize) / ChunkSize; nComplete >= 1 {
			k := min(1+nComplete, 8)
			if decrypt {
				fusedChunks = decryptChunk0Fused(&g, src[:k*ChunkSize], dst[:k*ChunkSize], k)
			} else {
				fusedChunks = encryptChunk0Fused(&g, src[:k*ChunkSize], dst[:k*ChunkSize], k)
			}
			src, dst = src[fusedChunks*ChunkSize:], dst[fusedChunks*ChunkSize:]
		}
		if fusedChunks == 0 {
			// Chunk 0: the trunk message phase, closed with MSG_LAST.
			g.trunk.bodyMore(dst[:ChunkSize], src[:ChunkSize], decrypt, msgMore)
			g.trunk.closeBlock(msgLast)
			src, dst = src[ChunkSize:], dst[ChunkSize:]
		}

		// Remaining complete leaf chunks, via the SIMD cascade.
		if n := len(src) / ChunkSize; n > 0 {
			g.processComplete(dst[:n*ChunkSize], src[:n*ChunkSize], n)
			src, dst = src[n*ChunkSize:], dst[n*ChunkSize:]
		}

		// Ragged tail: a final partial leaf of 1..ChunkSize-1 bytes.
		if len(src) > 0 {
			var leaf duplex
			if decrypt {
				decryptX1(g.key[:], g.nonce[:], g.nLeaves+1, src, dst, &leaf)
			} else {
				encryptX1(g.key[:], g.nonce[:], g.nLeaves+1, src, dst, &leaf)
			}
			t := leaf.tagBytes()
			g.trunk.absorbMore(t[:], aggMore)
			g.nLeaves++
		}
	}

	// Aggregation phase: leaf tags were absorbed in leaf order; append the
	// little-endian leaf count and close with AGG_LAST. The closing block's
	// keystream prefix is the root tag.
	var cnt [8]byte
	binary.LittleEndian.PutUint64(cnt[:], g.nLeaves)
	g.trunk.absorbMore(cnt[:], aggMore)
	g.trunk.closeBlock(aggLast)

	var tag [TagSize]byte
	g.trunk.extractTag(&tag)
	return tag
}

// processComplete processes nFlush complete leaf chunks via the SIMD cascade,
// absorbing their tags into the trunk in leaf order.
func (g *aggregator) processComplete(dst, src []byte, nFlush int) {
	idx := 0

	var tags [256]byte
	for idx+8 <= nFlush {
		off := idx * ChunkSize
		if g.decrypt {
			decryptChunks(g.key[:], g.nonce[:], g.nLeaves+1, src[off:off+8*ChunkSize], dst[off:off+8*ChunkSize], &tags)
		} else {
			encryptChunks(g.key[:], g.nonce[:], g.nLeaves+1, src[off:off+8*ChunkSize], dst[off:off+8*ChunkSize], &tags)
		}
		g.trunk.absorbMore(tags[:], aggMore)
		g.nLeaves += 8
		idx += 8
	}

	// 2-wide pass: drain the remaining complete chunks in pairs where a 2-wide
	// kernel is available (arm64). This avoids the x1 serial penalty for small
	// remainders (e.g. the three leaf chunks of a 32 KiB message). On
	// platforms without a 2-wide kernel the calls report false and the
	// register-resident and x1 passes below handle the remainder.
	for idx+2 <= nFlush {
		off := idx * ChunkSize
		var ok bool
		if g.decrypt {
			ok = decryptChunkPair(g, src[off:off+2*ChunkSize], dst[off:off+2*ChunkSize])
		} else {
			ok = encryptChunkPair(g, src[off:off+2*ChunkSize], dst[off:off+2*ChunkSize])
		}
		if !ok {
			break
		}
		idx += 2
	}

	// Remainder pass: drain a 2..7 chunk remainder with a single kernel call
	// that reads the chunks directly (no scratch buffer): register-resident
	// masked gather/scatter on AVX-512, dummy-lane x4 on AVX2. This removes
	// the x1 serial penalty for a small remainder (e.g. the three leaf chunks
	// of a 32 KiB message).
	//
	// This pass also takes the tail of a larger message (idx > 0), whose
	// chunks are cold: the kernels read each chunk as a sequential stream, so
	// cold tails prefetch as well as the x8 batches do. Measured on Emerald
	// Rapids (Xeon Platinum 8581C), the AVX-512 kernel beats both a padded-x8
	// pass and serial x1 for every remainder size, on cache-resident and cold
	// chunks alike; see the commit introducing this pass for the measurements.
	// On arm64 the pair pass already left fewer than two chunks; on platforms
	// without a remainder kernel the call reports false and the x1 loop below
	// runs.
	if rem := nFlush - idx; rem >= 2 {
		off := idx * ChunkSize
		var ok bool
		if g.decrypt {
			ok = decryptChunkRun(g, src[off:off+rem*ChunkSize], dst[off:off+rem*ChunkSize], rem)
		} else {
			ok = encryptChunkRun(g, src[off:off+rem*ChunkSize], dst[off:off+rem*ChunkSize], rem)
		}
		if ok {
			idx += rem
		}
	}

	// Remainder via x1: a single leftover chunk, or any remainder on platforms
	// without SIMD chunk kernels — where padding to 8 wide would buy nothing,
	// since the generic 8-way permute is eight serial permutes.
	var leaf duplex
	for idx < nFlush {
		off := idx * ChunkSize
		if g.decrypt {
			decryptX1(g.key[:], g.nonce[:], g.nLeaves+1, src[off:off+ChunkSize], dst[off:off+ChunkSize], &leaf)
		} else {
			encryptX1(g.key[:], g.nonce[:], g.nLeaves+1, src[off:off+ChunkSize], dst[off:off+ChunkSize], &leaf)
		}
		tag := leaf.tagBytes()
		g.trunk.absorbMore(tag[:], aggMore)
		g.nLeaves++
		idx++
	}
}

func decryptX1(key, nonce []byte, index uint64, ct, pt []byte, d *duplex) {
	p := leafInit(key, nonce, index)
	d.initWith(p[:])
	d.bodyMore(pt, ct, true, msgMore)
	d.closeBlock(msgLast)
}

func encryptX1(key, nonce []byte, index uint64, pt, ct []byte, d *duplex) {
	p := leafInit(key, nonce, index)
	d.initWith(p[:])
	d.bodyMore(ct, pt, false, msgMore)
	d.closeBlock(msgLast)
}

func checkSize(name string, got []byte, want int) {
	if len(got) != want {
		panic("tw128: invalid " + name + " size")
	}
}
